README.md
# Reference Manager

Emacs plugin to generate bibliography file automatically from a given plugin using various other libraries.
Also added functionality for browing gscholar with *eww*.

# Features

This project has now grown into a monolith and I'm refactoring it slowly to
decouple the features. These features are useful but not entirely stable at
present.

## Bibliography and file management

- Conversion from and to `org` property drawer to `bibtex` format
  - Import and export from `.bib` files
  - Sanitization and auto generation of keys
- Fetching of bibliography data from multiple sources and storing in org entries
  1. DBLP
  2. Semantic Scholar
  3. Crossref
  4. Google Scholar
- Fetching of pdfs from supported sources and storage in a dedicated directory
  See `ref-man-url-get-pdf-link-helper` in `ref-man-url.el`.

I've used a separate python module for network access with threading as it's
more efficient to do so. The communication is done via an http server.

## Reading and Navigation

Org headings serve as publication titles. For any notes one can simply store
them with the corresponding org headings. However for linking and citations
there are also some utility functions:

- Parsing of org headlines with filters in specified buffers into a cache.
- Insertion of link to any heading with `ido`
- Search and removal of duplicate headings.

## Research sources and Exploration

Currently two sources for exploration, search and archival are supported.

### Google Scholar

One can browse Google Scholar in a dedicated buffer derived from *eww* with
custom functions and keybindings for:

1. Easy navigation, filtering by date.
2. Import google scholar entry at point to org headline with metadata.
   - With optional fetching of PDF simultaneously
3. Import bibtex for google scholar entry at point.
4. Rendering via *chromium* debugger to avoid "prove you're not a robot".

### Semantic Scholar

We can also use [Semantic Scholar](https://www.semanticscholar.org). The module
can search Semantic Scholar with a search string, or lookup the SS database
through their API.

1. Search in Semantic Scholar and insert entry as org headline
2. Lookup an entry in Semantic Scholar database from any of the supported lookup
   types (arxiv, doi etc.)
3. Fetch the entry metadata and store in a cache
4. Parse the metadata and display in a separate org buffer with more details:
   - Abstract
   - Other Semantic Scholar metrics like `isInfluential`
   - All the parsed references from the metadata
   - All the papers which have cited that particular paper

Since we can see all the references and citations of that paper in a single org
buffer, we can view/download any references from the paper with ease without
constantly having to go to the back of the paper. We can also fetch the PDF (if
the source is supported by the module) and quickly check that paper also.

In addition, with all the papers which have cited the paper one is reading also
present, a quick bird's eye view of the state of the art is possible and more
recent interesting publications can also be downloaded quickly.

#### Science Parse

Semantic Scholar uses [science parse](https://github.com/allenai/science-parse)
to parse the PDFs' metadata. They provide the full model on that link and one
can also run that service locally in case one comes across a pdf not in their
database.

### Document preparation and publishing

Since we can embed Math markup, images, tables and links in the org buffer, it
can be exported to a fairly functional document. I've used a pandoc backend for
easy export to multiple formats.

#### Features

1. Automatic conversion of links in text to citations
   - A bibliography section is added automatically at the end if required.
2. Custom flags and switches for pandoc via its yaml header
3. Automatic insertion of additional bibliography files with yaml metadata
4. Easy export to html, PDF or Latex format

### Backup and File Sharing

The entire pdf and metadata cache can be uploaded to a supported cloud storage
for easy backup, access and sharing. I've used `rclone` for that and any backend
supported by `rclone` can therefore be theoretically used. We can:

1. Convert to html and mail an org buffer or subtree with pdf links being
   attached automatically from the specific remote storage.

For mail I use [mu4e](https://www.djcbsoftware.nl/code/mu/mu4e.html) and a
separate module which I call
[org-mu4e](https://github.com/akshaybadola/org-mu4e) which is built on top of
[org-mime](https://github.com/org-mime/org-mime) as a backend.

# Roadmap

There are some bugs and a lot of incomplete features. I had constructed a PyQt
GUI for viewing the citations as a graph but that project was shelved due to
lack of time. It can easily be repurposed and integrated with this project as a
backend.

Another very useful thing would be to have a JS based UI layer which can
interact with Emacs as a daemon for people who aren't so comfortable with
Emacs. We can parse `org` metadata (possibly with multiple threads) and render
it with HTML. It would be much more useful to the broader scientific community.

- Refactoring to make it more modular and remove redundant code.
- More comprehensive Documentation and Tutorial
- Unit/Regression testing setup
- Finish pending/incomplete features
- A mind-map/network layer for visualization
- UI layer on top for non emacs users as an optional module

# License

All the code in the repo is licensed under GPLv3. See LICENSE.md file in the repo.

For all libraries being used along with this codebase, please refer to their
licencses.

For any external modules or services (like Semantic Scholar or DBLP) being used,
please see their individual terms of services.



ChangeLog.md
# ChangeLog for the ref-man

## [2021-02-02 Tue 10:32]
- Fixed `ref-man--num-to-months`
- Added `condition-case` to `ref-man--validate-author`
- Changes to `ref-man-org-purge-entry`
- Added error message to `ref-man-parse-bib-property-key`
- Fixed extra `.pdf` being appended to filenames for `arxiv.org` urls
- Began implementing `ref-man-web-search`
- Some changes to `ref-man-web-gscholar`
- Added updating semantic scholar params with a node subprocess
- Added `progress` and `check_proxies` routes. `progress` is still pending. I
  think for that the Get class I'm trying to implement.

## [2021-02-11 Thu 02:52]
- Added .gitignore
- Version bump to 0.3.2
- Added url extraction from semanticscholar.org. Earlier it defaulted to first
  pdf link from buffer.
- Added `ref-man-url-get-best-pdf-url`.
- Lisp and python code follow separate versioning now.
- Fixed naming of PDF files downloaded from acm.org
- Fixed some pdf files names and downloads from openreview.net
- Fixed proxy_everything port.

## [2021-03-03 Wed 08:05]
- Python virtualenv is now always created and used.
- Staged debug_ss.js and package.json for chrome-remote
- Fixed python module `python -m ref_man` now runs correctly.

## [2021-03-04 Thu 16:28]
- `cache.py` is separate now and contains `CacheHelper`
- Pdfs remote links cache is now checked and fixed at startup if remote cache
  exists.

## [2021-03-06 Sat 00:36]
- Fixed some bugs in cache.
- Changed python process launch args to specify cache related args.
- Fixed an issue with semantic scholar params where it wouldn't switch to
  default if chrome was running and params couldn't be updated.

## [2021-03-22 Mon 08:56]
- Added `ref-man--invert-accents`.
- Strings are now inserted in org-mode without escaping non-ascii due to sytnax
  issues arising from that.
- Fixed a `cons` bug in `ref-man-url-get-best-pdf-url`
- Modified `ref-man-org-bibtex-read-from-headline`.
- Added `ref-man-org-get-bib-from-org-link`.
- Added command `ref-man-org-export-article` which exports the subtree to either pdf
  or html via pandoc. Primary advantage being auto generation of references.
- Defined `ref-man-bib-files` for `ref-man-org-export-article`. Can be added to
  other bibliography functions.

## [2021-03-23 Tue 08:38]
- Added `ref-man-fix-drawers-deleted-files` and utility function
  `ref-man-not-pdf-files`.

## [2021-03-23 Tue 08:38]
- `ref-man-chrome.el` is more modular now and later can be separated as a chrome
  debugger interface.
- New file `ref-man-py.el` separating python functionality from
  `ref-man-core.el`.
- Fix to `ref-man--generate-entry-from-hash` where I was passing `cons` instead
  of string to `ref-man--build-bib-key-from-plist`.
- Cleaner `ref-man--generate-buffer-and-fetch-if-required`
- Removed `util` as a dependency and added `ref-man-pairs-to-alist` which is a
  copy of `util/pairs-to-alist`.
- Some cleanup of the python module.
- Added `common_pyutil` as a dependency for the python module.
- Minor version bump to `0.4.0` with separation of python module.

## [2021-08-02 Mon 15:03]
- Modified README
- Updated requirements
- Version bump to `0.5.0`
- Replaced `cdass` with `a-get`
- Added the blog generation backend and settings for that.
- Added `ref-man-key-from-url` and `allow-misc` flag while exporting bibtex from
  org heading.
- Modularized `ref-man-docproc-export-article` and fixed some issues
- Fixed `References` heading was being added each time even without any
  citations
- Bibliography not in external files is now embedded in the yaml header itself.

## [2021-08-02 Mon 15:03]
- Added `ref-man-org-consolidate-drawers`, `ref-man-org-end-of-meta-data`,
  `ref-man-org-text-bounds`, `ref-man-org-bibtex-kill-headline-as-yaml`,
  `ref-man-org-delete-file-under-point`, `ref-man-org-update-from-from-crossref`,
  `ref-man-maybe-set-arxiv-id` in `ref-man-core.el`.
- Export functions are in new file `ref-man-export.el`.
- Added python version check in `ref-man-py-setup-env`
- Added fetcher for new version of AAAI site in
  `ref-man-url-get-pdf-link-helper` and some other additions.
- Added `ref-man-url-parse-cvf-venue` and `ref-man-url-cvf-pdf-link-helper`
  in `ref-man-url.el`.
- Changed Semantic Scholar function names, fetch and update are different
  functions now and you can display selective information.
- Fixed incorrect abstract formatting in `ref-man-org-insert-abstract` and it
  can optionally keep current text in entry
- Cleaned up initial server invocation from `ref_man.__main__.py`
- Changed `/get_cvpr_url` to `/get_cvf_url` and fixed some issues
- Added a lot of type annotations.
- Version bump to `0.6.0`.
- Py mod version bump to `0.3.2`.

## [2021-09-13 Mon 10:02]
- Bug fixes for regressions in ref-man-py.el and in cache.py
- Fixed `(message )` exprs.
- Separated `ref_man/util.py`
- Added more types in the python module.

## [2021-09-14 Tue 13:42]
- Added a `ref-man-export-python-executable`. Fixed a couple of issues with
  the customization executable and pandoc filters being in there.
- Fixed `ref-man-org-get-bib-from-org-link` error if bib cannot be parsed. Now
  it returns nil even if `get-key` is `t`. Earlier return '(path nil nil).
- Fixed `ref-man-docproc-export-article` including [@nil] for bib when unable to
  parse heading.
- Added `citeseerx.ist.psu.edu` as a pdf source
- Added function `sync_from_remote` in `ref_man.cache`.
- Version bump to `0.6.4`
- Py mod version bump to `0.3.4`

## [2021-09-21 Tue 03:14]
- Removed `ref-man-bibtex-ascii-accents` and separated
  `ref-man-bibtex-non-invertible-ascii-replacements` from
  `ref-man-bibtex-ascii-replacement-strings`.
- Cleaned up `ref-man--transcribe` a bit.
- Wrote utility functions `ref-man-org-bibtex-transform-author-bib-to-org`
  and `ref-man-org-bibtex-transform-author-org-to-bib`.
- Removed obsolete proxychains string.
- Replaced README.md with README.org
- Version bump from now on will be automatic with every push.

## [2021-09-24 Fri 12:22]

ref-man-py.el
;;; ref-man-py.el --- Module for managing the python process. ;;; -*- lexical-binding: t; -*-

;; Copyright (C) 2018,2019,2020,2021
;; Akshay Badola

;; Author:	Akshay Badola <akshay.badola.cs@gmail.com>
;; Maintainer:	Akshay Badola <akshay.badola.cs@gmail.com>
;; Time-stamp:	<Thursday 09 September 2021 01:23:15 AM IST>
;; Keywords:	pdfs, references, bibtex, org, eww

;; This file is *NOT* part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify it
;; under the terms of the GNU General Public License as published by the Free
;; Software Foundation; either version 3, or (at your option) any later
;; version.

;; This program is distributed in the hope that it will be useful, but WITHOUT
;; ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
;; FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
;; more details.

;; You should have received a copy of the GNU General Public License along with
;; GNU Emacs; see the file COPYING.  If not, write to the Free Software
;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

;;; Commentary:
;;
;; Functions for managing python virtualenv and processes for `ref-man'.  Has
;; functions to check if a virtualenv exists for a given path, if correct python
;; version is installed in the system and starting and stopping python
;; processes.

;;; Code:

(defcustom ref-man-py-executable nil
  "Path to python executable to use for `ref-man'."
  :type 'file
  :group 'ref-man)

(defcustom ref-man-py-server-port-start 9999
  "Server port on which to communicate with python server."
  :type 'integer
  :group 'ref-man)

(defcustom ref-man-py-server-port 9999
  "Port on which to communicate with python server."
  :type 'integer
  :group 'ref-man)

(defcustom ref-man-py-data-dir (expand-file-name "~/.ref-man/data/")
  "Server port on which to communicate with python server."
  :type 'directory
  :group 'ref-man)

(defcustom ref-man-proxy-port nil
  "Whether to use http proxy for all python requests.
If this is non-nil then the all the requests by the python server
go through this http proxy at localhost, specified by this port."
  :type 'integer
  :group 'ref-man)

(defcustom ref-man-chrome-debug-script (path-join (f-dirname (f-this-file)) "debug_ss.js")
  "Path to the chrome debugger javascript file.
The file contains code to get the Semantic Scholar Search
params.  As they can change, we need to update them once the
server starts.  Requires some kind of chrome(ium) to be installed
on the system."
  :type 'file
  :group 'ref-man)

;; NOTE: External variables
(defvar ref-man-home-dir)               ; from `ref-man'
(defvar ref-man-pdf-proxy-port)         ; from `ref-man-url'
(defvar ref-man-documents-dir)          ; from `ref-man-files'
(defvar ref-man-remote-documents-dir)   ; from `ref-man-remote'
(defvar ref-man-public-links-cache-file) ; from `ref-man-remote'

;; NOTE: Internal variables
(defvar ref-man-py-external-process-pid nil)

(defun ref-man-py-create-venv (python path)
  "Create a new `ref-man' virtual env in directory PATH.
The python executable to use is defined by PYTHON."
  (if (and (string-match-p "no.*module.*virtualenv.*"
                           (shell-command-to-string
                            (format "%s -m virtualenv -p %s %s"
                                    python (f-filename (f-canonical python)) path)))
           (string-match-p "no.*module.*virtualenv.*"
                           (shell-command-to-string
                            (format "%s -m virtualenv -p %s %s"
                                    python (f-filename (f-canonical python)) path))))
      nil (message "Create venv with %s in %s" (f-filename (f-canonical python)) path)))

(defun ref-man-py-no-mod-in-venv-p (python)
  "Check if `ref-man' python environment has py module installed.
PYTHON is the python executable for that virtualenv."
  (string-match-p "no.*module.*ref.*"
                  (shell-command-to-string
                   (format "%s -m ref_man --version" python))))

(defun ref-man-py-installed-mod-version (python)
  "Return version for installed python module.
PYTHON is the path for python executable."
  (shell-command-to-string
   (format "%s -m ref_man --version" python)))

(defun ref-man-py-file-mod-version ()
  "Return the version of python module in file `const.py'."
  (let* ((const-file (path-join ref-man-home-dir "ref_man" "const.py"))
        (buf (find-file-noselect const-file)))
    (with-current-buffer buf
      (goto-char (point-min))
      (re-search-forward "__version__\\(.+?\\)=\\(.+?\\)\"\\(.+\\)\"")
      (substring-no-properties (match-string 3)))))

(defun ref-man-py-venv-python-version ()
  "Return version of python executable in `ref-man' virtualenv."
  (cadr (split-string
         (shell-command-to-string
          (concat (path-join ref-man-home-dir "env" "bin" "python") " " "--version")))))

(defun ref-man-py-env-needs-update-p ()
  "Check if `ref-man' module needs to be updated."
  (not (equal (ref-man-py-file-mod-version)
              (string-trim (replace-regexp-in-string
                            "ref-man.*? version \\(.*\\)" "\\1"
                            (ref-man-py-installed-mod-version
                             (path-join ref-man-home-dir "env" "bin" "python")))))))

(defun ref-man-py-env-uninstall-module (env)
  "Install the `ref-man' module in virtualenv ENV."
  (shell-command
   (concat "source " (path-join env "bin" "activate") " && pip uninstall -y ref-man")
   "*ref-man-uninstall-cmd*" "*ref-man-uninstall-cmd*"))

(defun ref-man-py-env-install-module (env)
  "Uninstall the `ref-man' module from virtualenv ENV."
  (shell-command
   (concat "source " (path-join env "bin" "activate") " && "
           (format "cd %s && python -m pip install -U ."
                   ref-man-home-dir))
   "*ref-man-install-cmd*" "*ref-man-install-cmd*"))

(defun ref-man-py-get-system-python ()
  "Get system python3 executable."
  (unless (or ref-man-py-executable (executable-find "python3"))
    (error "No `ref-man-py-executable' or python3 found in current paths.\n
Either set `ref-man-py-executable' or add python3 path to the exec path"))
  (or ref-man-py-executable (executable-find "python3")))

(defun ref-man-py-setup-env (&optional reinstall update)
  "Setup python virtualenv.

Optional non-nil REINSTALL removes the virtualenv and installs
everything again.  Optional non-nil UPDATE only updates the
`ref-man' python module.  The directory is relative to `ref-man'
install directory `ref-man-home-dir'."
  (let* ((python (ref-man-py-get-system-python))
         (py-version (and python (cadr (split-string (shell-command-to-string (format "%s --version" python))))))
         (env (path-join ref-man-home-dir "env")))
    (when (version< py-version "3.6.9")
      (user-error "Your default python3 executable is < 3.6.9.
`ref-man' requires at least 3.6.9.
Please set `ref-man-py-executable' with a version of python >= 3.6.9"))
    (when (and reinstall (f-exists? env)
               (y-or-n-p (format "Clean and reinstall virtualenv %s? " env)))
      (f-delete env t))
    (unless (f-exists? env)
      (f-mkdir env)
      (unless (ref-man-py-create-venv python env)
        (error "Could not install venv.\n
Make sure package 'virtualenv' exists in current python environment")))
    (let* ((env-has-python (f-exists? (path-join env "bin" "python3")))
           (python (and env-has-python
                        (path-join env "bin" "python3"))))
      (when (and (not python) (ref-man-py-create-venv python env))
        (error "Could not install venv.\n
Make sure package 'virtualenv' exists in current python environment"))
      (let* ((env-has-no-ref-man (ref-man-py-no-mod-in-venv-p python))
             (need-update (or update (ref-man-py-env-needs-update-p))))
        (when (or need-update env-has-no-ref-man)
          (cond (need-update
                 (message "New version of ref-man. Updating existing ref-man-server in %s" env)
                 (ref-man-py-env-uninstall-module env))
                (env-has-no-ref-man
                 (message "ref-man-server not found. Installing in %s" env)))
          (ref-man-py-env-install-module env))
        (if (ref-man-py-installed-mod-version python)
            (message "%s found in %s" (ref-man--trim-whitespace
                                       (ref-man-py-installed-mod-version python))
                     env)
          (error "Could not install ref-man in %s" env))))))

;; TODO: Requests to python server should be dynamic according to whether I want
;;       to use proxy or not at that point
(defun ref-man-py-process-helper (data-dir port &optional proxy-port docs-dir)
  "Start the python server.
DATA-DIR is the server data directory.  PORT is the port to which
the server binds.

When called from `ref-man-py-start-server', DATA-DIR is set
to `ref-man-py-data-dir' and the port
`ref-man-py-server-port'.

Optional arguments PROXY-PORT, PROXY-EVERYTHING-PORT, DOCS-DIR
are specified for modularity and are set to `ref-man-proxy-port',
nil, `ref-man-py-data-dir' respectively by
`ref-man-py-start-server'."
  ;; NOTE: Hack so that process isn't returned
  (prog1
      (message (format "[ref-man] Starting python process on port: %s"
                       ref-man-py-server-port))
    (let ((python (path-join ref-man-home-dir "env" "bin" "python"))
          (proxy-port (or proxy-port ref-man-proxy-port)))
      (condition-case nil
          (ref-man-py-setup-env))
      (let ((args (-filter #'identity (list (format "--data-dir=%s" data-dir)
                                            (format "--port=%s" port)
                                            (and proxy-port "--proxy-everything")
                                            (and proxy-port
                                                 (format "--proxy-everything-port=%s" proxy-port))
                                            (and ref-man-pdf-proxy-port
                                                 (format "--proxy-port=%s"
                                                         ref-man-pdf-proxy-port))
                                            (and ref-man-chrome-debug-script
                                                 (format "--chrome-debugger-path=%s"
                                                         ref-man-chrome-debug-script))
                                            (and ref-man-documents-dir
                                                 (format "--local-pdfs-dir=%s"
                                                         ref-man-documents-dir))
                                            (and ref-man-remote-documents-dir
                                                 (format "--remote-pdfs-dir=%s"
                                                         ref-man-remote-documents-dir))
                                            (and ref-man-public-links-cache-file
                                                 (format "--remote-links-cache=%s"
                                                         ref-man-public-links-cache-file))
                                            "--verbosity=debug"))))
        (message "Python process args are %s" args)
        (apply #'start-process "ref-man-server" "*ref-man-server*"
               python "-m" "ref_man" args)))))

(defun ref-man-py-stop-server ()
  "Stop the python server by sending a shutdown command.
This is sent via http and lets the server exit gracefully."
  (interactive)
  (let ((buf (url-retrieve-synchronously
              (format "http://localhost:%s/shutdown" ref-man-py-server-port))))
    (with-current-buffer buf
      (goto-char (point-min))
      (re-search-forward "\r?\n\r?\n")
      (message (buffer-substring-no-properties (point) (point-max))))))

(defun ref-man-py-kill-internal-process ()
  "Kill the internal python process process by sending SIGKILL."
  (signal-process (get-buffer "*ref-man-server*") 15))

(defun ref-man-py-kill-external-process ()
  "Kill the external python process process by sending SIGKILL."
  (signal-process ref-man-py-external-process-pid 15))

(defun ref-man-py-server-reachable-p ()
  "Check if python server is reachable."
  (condition-case nil
      (let ((buf (url-retrieve-synchronously
                  (format "http://localhost:%s/version" ref-man-py-server-port) t)))
        (when buf
          (string-match-p "ref-man python server"
                          (with-current-buffer buf (buffer-string)))))
    (error nil)))

(defun ref-man-py-process-running ()
  "Check if python server is running.
Returns 'external or 'internal according to where the process is
running if it's running else nil."
  (cond ((get-buffer-process "*ref-man-server*")
         (setq ref-man-py-external-process-pid nil)
         'internal)
        ((ref-man-py-external-process-p)
         'external)
        (t nil)))

(defun ref-man-py-server-running ()
  "Check if python server is already running."
  (let ((python-process (ref-man-py-process-running)))
    (when python-process
      (if (ref-man-py-server-reachable-p)
          python-process
        (if (eq python-process 'internal)
            'internal-error 'external-error)))))

(defun ref-man-py-external-process-p ()
  "Check for `server.py' python processes outside Emacs.
In case a process is found, `ref-man-py-server-port' is set
to the port of that process and
`ref-man-py-external-process-pid' is set to its pid."
  (let ((python-strings
         (split-string (shell-command-to-string "ps -ef | grep python | grep server") "\n")))
    (cl-loop for x in python-strings
             do
             (when (and (string-match-p "port" x) (string-match-p "data-dir" x))
               (setq ref-man-py-server-port
                     (string-to-number
                      (cadr (split-string
                             (car (split-string
                                   (substring x (string-match "port" x)))) "="))))
               (setq ref-man-py-external-process-pid (string-to-number (nth 1 (split-string x))))
               (cl-return t)))))

(defun ref-man-py-start-server ()
  "Start the python server, unless already running.

The server can be running outside Emacs also in which case
`ref-man-py-server-port' is set to port.

See accompanying `server.py' for the server details.  The API and
methods are still evolving but as of now it supports DBLP and
ArXiv.  The process if started opens a local port and can fetch
data in multiple threads from supported APIs before preprocessing
and consolidating.  It also maintains a local datastore."
  (interactive)
  ;; FIXME: for 'internal-error and 'external-error
  (if (ref-man-py-server-running)
      (message (format "Found existing process running on port: %s"
                       ref-man-py-server-port))
    (message "No existing python process found")
    (let ((port (find-open-port ref-man-py-server-port-start))
          (data-dir ref-man-py-data-dir))
      (setq ref-man-py-server-port port)
      (ref-man-py-process-helper data-dir port))))

(defun ref-man-py-restart-server ()
  "Restart the python server."
  (interactive)
  (cond ((or (eq 'internal (ref-man-py-server-running))
             (eq 'external (ref-man-py-server-running)))
         (ref-man-py-stop-server))
        ((eq 'internal-error (ref-man-py-server-running))
         (ref-man-py-kill-internal-process))
        ((eq 'external-error (ref-man-py-server-running))
         (ref-man-py-kill-external-process)))
  ;; FIXME: This runs before server shuts down
  (ref-man-py-start-server))

(provide 'ref-man-py)

;;; ref-man-py.el ends here
